Update update_fields arg in custom save methods if specified

gherceg opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->

## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
https://dimagi.atlassian.net/browse/SAAS-15436

In reviewing changes in Django 4.2, I came across [this one](https://docs.djangoproject.com/en/5.0/releases/4.2/#setting-update-fields-in-model-save-may-now-be-required)

> In order to avoid updating unnecessary columns, [QuerySet.update_or_create()](https://docs.djangoproject.com/en/5.0/ref/models/querysets/#django.db.models.query.QuerySet.update_or_create) now passes update_fields to the [Model.save()](https://docs.djangoproject.com/en/5.0/ref/models/instances/#django.db.models.Model.save) calls. As a consequence, any fields modified in the custom save() methods should be added to the update_fields keyword argument before calling super()

This actually didn't immediately apply to us as I didn't find any current uses of `update_or_create` on models with custom saves that modify fields. However to future proof against what I think would be unexpected behavior if using `update_or_create` on a model that does have a custom save method, I've identified the current custom save methods that modify fields and therefore should update the `update_fields` kwarg if specified.

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
Realistically this should have no impact on existing code since there aren't places where we pass `update_fields` into these custom save methods. 
### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
No.

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

gherceg committed:
Update update_fields if supplied in kwargs in locations app

- LocationType
- SQLLocation

gherceg committed:
Update update_fields if supplied in kwargs in users app

gherceg committed:
Update update_fields if supplied in kwargs in generic_inbound app

gherceg committed:
Update update_fields if supplied in kwargs in repeaters app

kaapstorm reviewed (approved):
I think this is fine. In the case of `HQApiKey` and `ConfigurableAPI` the change can't be reached, because it's nested inside an `if` clause that is only true on create, not update. For the commtrack/location change, I'm a little disappointed (but not really surprised :thinking: ) that this change didn't affect a single test, but writing new tests for this feels out of scope, considering this is just to bring our code in line with an upstream library update.

gherceg commented:
> In the case of HQApiKey and ConfigurableAPI the change can't be reached, because it's nested inside an if clause that is only true on create, not update

Well the code would be reached on create, which is possible via `update_or_create` right? And if it is just an update, then it is good that this code isn't reached since it doesn't apply at that point? Am I misinterpreting what you said? Or maybe even the feature change?

gherceg commented:
Offlined with @kaapstorm and he explained to me that `update_fields` is likely unused on create (of course), which is why the code path isn't relevant.

millerdev commented on corehq/apps/commtrack/models.py:
```diff
@@ -292,6 +292,9 @@ def sync_supply_point(location, is_deletion=False):
         updated_supply_point = _reopen_or_create_supply_point(location)
         location.supply_point_id = updated_supply_point.case_id
 
+    if update_fields:
```
I think this and other similar conditions in this PR should check if `is not None` since I think `update_fields` could be an empty list, and this condition should always be triggered if it's a list.

gherceg replied:
Good catch. Resolved in 7d63f0f856fc8bb92521e3efa0d7259458b06bd7

millerdev commented on corehq/apps/locations/models.py:
```diff
@@ -157,23 +157,32 @@ def expand_from_root(self, value):
+            self._populate_stock_levels(config, update_fields=additional_update_fields)
 
         is_not_first_save = self.pk is not None
+        if 'update_fields' in kwargs.keys():
```
nit
```suggestion
        if 'update_fields' in kwargs:
```
There are a bunch of other superfluous `.keys()` calls in this PR that could also be changed in the same way.

gherceg replied:
Thanks. See 7d63f0f856fc8bb92521e3efa0d7259458b06bd7

gherceg committed:
Address PR feedback

- check that update_fields is not None since an empty list could be
  passed into the method
- replace superfluous use of kwargs.keys() with kwargs

millerdev commented on corehq/apps/locations/models.py:
```diff
@@ -759,6 +770,8 @@ def set_site_code_if_needed(location):
                                 .values_list('site_code', flat=True))
         ]
         location.site_code = generate_code(location.name, all_codes)
+        if update_fields:
```
Missed one?

```suggestion
        if update_fields is not None:
```

gherceg replied:
😒 my quickfix list let me down. Thanks

gherceg committed:
Check that update_fields is not None since it could be an empty list

Missed one from my previous commit that was supposed to resolve this.

gherceg committed:
More thorough check that update_fields can and should be appended to

millerdev reviewed (commented):
I don't think this is necessary. It would be a bug if someone called `obj.save(update_fields=None)`, and I think we want to it raise an exception if that happens (and that would happen on `kwargs['update_fields'].extend(...)`) rather than allowing the error to pass silently.

Edit: this was only intended to refer to the [most recent commit](https://github.com/dimagi/commcare-hq/pull/34402/commits/c30091f07a58be936eaf21d495bcdc99ff55a6bd).

gherceg commented:
> It would be a bug if someone called obj.save(update_fields=None)

This makes sense. I'll revert that last commit.

gherceg committed:
Revert "More thorough check that update_fields can and should be appended to"

This reverts commit c30091f07a58be936eaf21d495bcdc99ff55a6bd.
